File: frontend/src/types/visitor.ts
export type VisitorInput = {
  name: string;
  surname: string;
  birth_date: Date | null;
  gender: string;
  amka: string;
  weight: string;
  height: string;
  smoker: string;
  years_smoking: string;
  cig_per_day: string;
  email: string;
  phoneNumber: string;
  history: string;
  medicines: any[];
  diseases: any[];
};

export type Visitor = {
  id: number;
  name: string;
  surname: string;
  amka: string;
  latest_visit: string | null;
};

export type VisitorsHomePage = {
  visitors: Visitor[];
  next_cursor: number;
  has_more: boolean;
};

export type VisitSummary = {
  id: number | null;
  diagnosis: string | null;
  created_at: string | null;
};

export type DiseaseRef = { id: number; name: string };
export type MedicineLink = { medicine: { id: number; name: string }; until: string | null };

export interface VisitorDetail {
  id: number | null;
  name: string | null;
  surname: string | null;
  amka: string | null;
  birth_date: string | null;
  gender: "male" | "female" | "other" | null;
  email: string | null;
  phone_number: string | null;
  smoker: "smoker" | "non_smoker" | "ex_smoker";
  years_smoking: number | null;
  cig_per_day: number | null;
  height: number | null;
  weight: number | null;
  history: string | null;
  diseases: DiseaseRef[];
  medicines_links: MedicineLink[];
  visits: VisitSummary[];
}

File: backend/schemas/visitor.py
import enum
from datetime import date,datetime
from typing import Optional,List
from pydantic import BaseModel,EmailStr
from .medicine import *
from .disease import *
from .visit import VisitResponseForVisitor

class GenderEnum(enum.Enum):
    male = "male"
    female ="female"
    other ="other"

class SmokerEnum(enum.Enum):
    smoker = "smoker"
    non_smoker="non_smoker"
    ex_smoker="ex_smoker"

class VisitorBase(BaseModel):
    name : str
    surname : str
    birth_date : date
    gender : GenderEnum
    amka : Optional[str] = None
    weight : int
    height : float
    smoker : SmokerEnum
    years_smoking:Optional[int] = None
    cig_per_day:Optional[int] = None

class VisitorCreate(VisitorBase):
    email : Optional[EmailStr] = None
    phone_number : Optional[str] = None
    history : Optional[str] = None


class VisitorMedicineLink(BaseModel):
    medicine:MedicineResponse
    until:Optional[date]

class VisitorResponse(VisitorBase):
    id:int
    email : Optional[EmailStr]
    phone_number : Optional[str]
    history : Optional[str]
    medicines_links : Optional[List[VisitorMedicineLink]]
    diseases : Optional[List[DiseaseResponse]]


class VisitorResponseWithVisits(VisitorResponse):
    visits : Optional[List[VisitResponseForVisitor]]


class VisitorHomepage(BaseModel):
    id : int
    name : str
    surname : str
    amka : Optional[str] 
    latest_visit : Optional[datetime] = None

    class Config:
        from_attributes = True

class VisitorResponseHomepage(BaseModel):
    visitors: List[VisitorHomepage]
    next_cursor:Optional[int]
    has_more:bool